xbmc/guilib/GUIComponent.cpp

```diff
@@ -40,16 +41,11 @@ void CGUIComponent::Init()
   m_stereoscopicsManager->Initialize();
   m_guiInfoManager->Initialize();
 
-  //! @todo This is something we need to change
-  m_pWindowManager->AddMsgTarget(m_stereoscopicsManager.get());
-
-  CServiceBroker::RegisterGUI(this);
+  AddMsgTarget(m_stereoscopicsManager.get());
 }
 
 void CGUIComponent::Deinit()
 {
-  CServiceBroker::UnregisterGUI();
```

Comment:
unregister from ServiceBroker is mandatory here. Deinit puts the service out of order

Reply:
But it’s inconsistent to have it done in the object that gets registered. The owner is application - the owner is responsible to resolve the injected dependencies - like it is done with other broker registrees already.

Reply:
you only unregister it at cleanup but it has to be unregistered during unload/load skin and maybe other operation too. encapsulation makes sense here.
btw: unregister needs to wait until last client has released the shared pointer. that's the reason why it is a shared pointer: https://github.com/xbmc/xbmc/blob/master/xbmc/Application.cpp#L2535

Reply:
Is this new behavior? Can't see where the GUI was deregistered during unload/load skin before tbh. But its a good idea to change skins only while no GUI Users are calling into the component for sure.

Reply:
I have added some more commits towards your directions. Though the check for the gui pointer might be not easily doable ... i started it but still have 1025 hits on unguarded accesses ... :(

Reply:
> Is this new behavior? Can't see where the GUI was deregistered during unload/load skin before tbh

Sorry, I forgot to mention that this was my personal view on this, not an official direction of team kodi.
The scenario I had in mind was i.e. that you run headless kodi, launch an X11 desktop and Kodi Ui gets visible, close the desktop -> Kodi falls back to headless, log on to Wayland -> you get Wayland UI of Kodi, ...

Reply:
As far as I am concerned nobody in team kodi has GUI ownership atm - so no need to stress that your opinion is not official or what not. Just discussing improvements here and trying to find a way to give that shared pointer a meaning ;)

Reply:
True and true.
We still want to support the scenario you mentioned @fernetmenta and appreciate your feedback a lot.
I think I speak for the team, when I say this.